cmd/observe/io_reader_observer.go

```diff
@@ -75,36 +80,109 @@ func newIOReaderClient(ctx context.Context, scanner *bufio.Scanner, request *obs
 	if err != nil {
 		return nil, err
 	}
+
+	var buf *container.RingBuffer
+	// last
+	if !request.GetFirst() && request.GetNumber() != 0 && request.GetNumber() != math.MaxUint64 {
+		buf = container.NewRingBuffer(int(request.GetNumber()))
+	}
```

Comment:
There is a whole range for `n` leading to a panic because of `int` overflow (at least `math.MaxInt64 < n < math.MaxUint64`) and OOM potential, so I think we should bail out if `n` is "too large". Assuming that a flow is ~5KiB in memory it already adds up to ~5GiB only for the ring buffer flows when  `n = 1_000_000`.

```suggestion
	// --last
	if n := request.GetNumber(); !request.GetFirst() && n != 0 && n != math.MaxUint64 {
		if n > 1_000_000 {
			return nil, fmt.Errorf("--last must be <= 1_000_000, got %d", n)
		}
		buf = container.NewRingBuffer(int(n))
	}
```

Reply:
This seems pretty reasonable, and we can always come back to relaxing the limit if it's an issue in practice.